### ISSUE TITLE
[no-namespace] [prefer-namespace-keyword] Default Settings Conflict

### ISSUE THREAD

**Issue by Deafwave** [typescript-eslint/typescript-eslint#324]
**Repro**

```JSON
{
    "parser": "@typescript-eslint/parser",
    "env": {
        "jest": true,
        "es6": true,
        "node": true
    },
    "extends": [
        "plugin:@typescript-eslint/recommended"
    ],
    "globals": {},
    "parserOptions": {},
    "plugins": [],
    "rules": {}
}
```

```TS
export module foo {
    export interface bar {

    }
}
```

**Expected Result**
Single route of optimizing code to new standards.

**Actual Result**
"Use 'namespace' instead of 'module' to declare custom TypeScript modules."
"ES2015 module syntax is preferred over custom TypeScript modules and namespaces."

These two rules contradict each other.

**Additional Info**
This may be due to human error, I am not very familiar with module syntax. My goal was to make a namespace with interfaces inside of it to utilize in my class. What's the most optimal way to do this? Export a module and import the module, then use that module in the class is the "new" scenario that I found based off of these rules.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `^1.4.2` |
| `@typescript-eslint/parser`        | `^1.4.2` |
| `TypeScript`                       | `^3.1.3` |
| `ESLint`                           | `^5.14.1` |
| `node`                             | `10.13.0` |
| `npm`                              | `6.4.1` |

**Comment by bradzacher** [comment]
The rules aren't conflicting, they're telling you two different ways to achieve this.


AFAIK modules are really only used now-a-days to define ambient typings for npm packages.
```ts
declare module 'typescript' {
    export function //...
}
```


Namespaces are a way to do this within your code (which is the first error), though I would suggest moving the interfaces into their own file (i.e. es modules, the second error).

consider
```TS
namespace Foo {
	interface Bar {}
	interface Baz {}
}

export class FooBar implements Foo.Bar, Foo.Baz {}
```

versus
```TS
// ifaces.ts
export interface Bar {}
export interface Baz {}

// class.ts
import * as Foo from './ifaces.ts'

export class FooBar implements Foo.Bar, Foo.Baz {}
```


Closing as not a bug.

**Comment by G-Rath** [comment]
@bradzacher would it be possible to have an example of this added to the rule?

I know there are resources out there that cover this, but I've not been able to find anything that explicitly says "TS decided to move away from namespaces & modules completely - instead use import/export, & here's what that looks like".

I'm in a situation where I have a tight deadline, so while I'd love to spend time reading the detailed & in-depth articles, I sadly don't have the time :( - it'd be great if there was just a code snippet on the `no-namespaces` rule showing a before & after :)

From my understanding, this is the before & after:

```
// definitions.ts
export namespace MyApi {
  export interface ApiResponse {
    entities: Array<any>;
  }
}
```

```
// MyApi.ts
export interface ApiResponse {
  entities: Array<any>;
}
```

Please correct me if I'm wrong :)

**Comment by bradzacher** [comment]
Don't take this as the complete truth, it's been a long time since this was the case, so I'm piecing it together based on what I remember......

Back in the day when TS was introduced, import/export syntax wasn't implemented (or even fully defined yet?), so you instead had to import types via the triple slash reference `/// <reference path="..." />`

What this would do was essentially import your types as if they were included in the file.

```TS
// types.ts
interface Foo {}
interface Bar {}


// code.ts
/// <reference path="./types" />

const foo: Foo = {};
const bar: Bar = {};
```

This is obviously a problem, because if you happen to define `interface Foo` in your local file, then there is clashing, and then type merging rules come into play. Similarly it can litter your local file's scope with types that you mightn't have wanted (which makes autocomplete/intellisense less useful).

So they added namespaces to allow you to, well, namespace your pseudo-imports
```TS
// types.ts
namespace Types {
  interface Foo {}
  interface Bar {}
}

// code.ts
/// <reference path="./types" />

const foo: Types.Foo = {};
const bar: Types.Bar = {};
```

Also important to note that around this time, module based bundlers like webpack were still in their early stages, so most people were using the very ludicrous old school way of compiling projects - concatenating every single file together. So namespaces also helped handle this pattern nicely by providing a nice syntax which [automatically transpiled to IIFE syntax](https://www.typescriptlang.org/play/#src=namespace%20Foo%20%7B%0A%20%20%20%20export%20function%20bar()%20%7B%0A%0A%20%20%20%20%7D%0A%7D%0A%0AFoo.bar()%3B).

Then import/export syntax came around, which enabled strictly typed module boundaries and allowed you to import/export types. So ambient rules no longer applied, and namespaces became pretty much redundant.
Additionally module bundlers matured and the ecosystem settled on the idea of modules (commonjs (aka what node uses), amd, etc), so the idea of using explicitly defined namespaces for your actual code fell out of fashion.

```TS
// types.ts
export interface Foo {}
export interface Bar {}

// code.ts
import * as Types from "./types"

const foo: Types.Foo = {};
const bar: Types.Bar = {};
```

Some people still use namespaces to help group things together (i.e. they might have a large number of different, semi-types within the same file, so grouping them makes it a bit more organised), but it's generally much preferred to just create completely separate files and have the module system do your grouping implicitly.

----

In the same vein, the `module "name" {}` syntax exists for one reason - to provide you with a way to limit the scope of ambient types.
Generally local `*.d.ts` files are included in the ambient scope at compile time, so by wrapping your types in the `module "name" {}` syntax, you can inform the compiler to ignore these types unless it sees `import X from "name"`.

Note that types from DefinitelyTyped (i.e. the `@types/name` packages) have some magic done on them via default config such that typescript uses the folder path to figure out the module name and essentially automatically wrap them in a module definition.

**Comment by kumarharsh** [comment]
Can't imagine I had to dig through to this issue to finally understand the problem. Thanks for the clear explanation @bradzacher!
